### PR TITLE
feat: support Sigv4 for non AWS services

### PR DESCRIPTION
**Pull request by gosar**
### Description
This has changes to `middleware-signing` and `codegen` to support generating clients for non AWS services that have SigV4 auth.

### Testing
I generated client for the bootleg demo service we are working on and with SigV4 enabled on the service, confirmed the client is able to successfully call it. See https://github.com/adamthom-amzn/smithy-typescript-ssdk-demo/pull/20 for testing details.

I ran `yarn generate-clients`. There is only a change is comment for all generated clients, but it is actually going back to what it was few weeks back before my earlier changes for supporting non AWS services changed the comment slightly.

---
By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

File: codegen/smithy-aws-typescript-codegen/src/main/java/software/amazon/smithy/aws/typescript/codegen/AddAwsRuntimeConfig.java [hunk 1]
```diff
@@ -95,8 +95,13 @@ public void addConfigInterfaceFields(
...
         if (isSigV4Service(settings, model)) {
-            writer.writeDocs("The AWS region to which this client will send requests or use as signingRegion")
-                    .write("region?: string | __Provider<string>;\n");
+            if (isAwsService(settings, model)) {
+                writer.writeDocs("The AWS region to which this client will send requests")
+                        .write("region?: string | __Provider<string>;\n");
+            } else {
+                writer.writeDocs("The AWS region to use as signing region for AWS Auth")
+                        .write("region?: string | __Provider<string>;\n");
+            }
```
Comment (JordonPhillips) [inline codegen/smithy-aws-typescript-codegen/src/main/java/software/amazon/smithy/aws/typescript/codegen/AddAwsRuntimeConfig.java:104]: Writing the actual region can be outside of this block since it doesn't change.

EDIT: actually it seems like the rest of the changes in this PR serve to effectively make region required, so I guess you'd want to enforce that here

File: packages/middleware-signing/src/configurations.ts [hunk 2]
```diff
@@ -32,6 +32,29 @@ export interface AwsAuthInputConfig {
    */
   signingRegion?: string;
 }
+
+export interface CustomAwsAuthInputConfig {
```
Comment (JordonPhillips) [inline packages/middleware-signing/src/configurations.ts:36]: Why is this needed?

File: codegen/smithy-aws-typescript-codegen/src/main/java/software/amazon/smithy/aws/typescript/codegen/EndpointGenerator.java [hunk 3]
```diff
@@ -58,6 +58,8 @@
         serviceTrait = service.getTrait(ServiceTrait.class)
                 .orElseThrow(() -> new CodegenException("No service trait found on " + service.getId()));
         endpointPrefix = serviceTrait.getEndpointPrefix();
+        // Isn't SigV4 trait required for AWS service and isn't SigV4Trait::getName required.
+        // So why the fallback to getArnNamespace?
```
Comment (JordonPhillips) [inline codegen/smithy-aws-typescript-codegen/src/main/java/software/amazon/smithy/aws/typescript/codegen/EndpointGenerator.java:62]: New auth types can always be added in the future

**Review by JordonPhillips — CHANGES_REQUESTED** [review]
Why is this bifurcation needed? It looks to me like all you're doing is avoiding the region provider, but should you?

File: packages/middleware-signing/src/configurations.ts (hunk 2 above)
Comment (gosar) [reply packages/middleware-signing/src/configurations.ts:36]: For non-AWS service, we don't need signingRegion as a config input. The region acts as the signingRegion.

File: codegen/smithy-aws-typescript-codegen/src/main/java/software/amazon/smithy/aws/typescript/codegen/EndpointGenerator.java (hunk 3 above)
Comment (gosar) [reply codegen/smithy-aws-typescript-codegen/src/main/java/software/amazon/smithy/aws/typescript/codegen/EndpointGenerator.java:62]: Yes, but how can you tell now what the fallback should be for future auth types? When the case arises, I'm sure we can deal with the changes then. It seems, I could be wrong and hence this is a question, the fallback is dead code.

File: codegen/smithy-aws-typescript-codegen/src/main/java/software/amazon/smithy/aws/typescript/codegen/AddAwsRuntimeConfig.java (hunk 1 above)
Comment (gosar) [reply codegen/smithy-aws-typescript-codegen/src/main/java/software/amazon/smithy/aws/typescript/codegen/AddAwsRuntimeConfig.java:104]: I had the `if` to write a specific comment in each case.

It is not required in the sense that you don't need to provide it when instantiating the client. It is link for [AWS clients](https://github.com/aws/aws-sdk-js-v3/blob/main/clients/client-accessanalyzer/AccessAnalyzerClient.ts#L231) where it comes in to [ClientDefaultValues](https://github.com/aws/aws-sdk-js-v3/blob/main/clients/client-accessanalyzer/AccessAnalyzerClient.ts#L310) from [here](https://github.com/aws/aws-sdk-js-v3/blob/main/clients/client-accessanalyzer/runtimeConfig.ts#L32)

**Comment by gosar** [comment]
> Why is this bifurcation needed? It looks to me like all you're doing is avoiding the region provider, but should you?
> 
> EDIT: I may be confusing two things. A region provider can be a thing that pulls the region from, say, environment variables or EC2 metadata. Or it could be the thing that uses the region to resolve an endpoint. The latter we definitely want to cut out, but the former should be kept if possible. If they're inextricably linked then I guess we have no choice, but it'd be good to leave a todo so we can explore splitting them up in the future.

I guess you are asking about bifurcation of AwsAuth v/s CustomAwsAuth. Yeah it's the latter. The [RegionInfoProvider](https://github.com/aws/aws-sdk-js-v3/blob/main/packages/types/src/util.ts#L83-L103) determines the endpoint and signingService, signingRegion.

File: codegen/smithy-aws-typescript-codegen/src/main/java/software/amazon/smithy/aws/typescript/codegen/EndpointGenerator.java (hunk 3 above)
Comment (gosar) [reply codegen/smithy-aws-typescript-codegen/src/main/java/software/amazon/smithy/aws/typescript/codegen/EndpointGenerator.java:62]: This comment is related to https://github.com/aws/aws-sdk-js-v3/pull/1835 which says _and fall back to arnNamespace(for non-AWS clients)_ , which I don't understand. Why would non-AWS clients have an arnNamespace? They wouldn't have the `ServiceTrait` at all.

**Comment by gosar** [comment]
Based on discussion with @AllanZhengYP,
> 1. Any alternate naming suggestions for CustomAwsAuth?

I'm going to call this SigV4Auth (for non-AWS services) and AwsAuth will represent SigV4 auth for AWS services.

> 2. There was existing reference to input.signingName in middleware-signing but I couldn’t see how one provides that input configuration. Is it safe to add signingName in config interface for every AWS service?

It seems safe to add signingName for every AWS service. I'll mark it `@internal`. The existing [logic](https://github.com/aws/aws-sdk-js-v3/blob/4c535c97d280e15bfc894737d7837fd85aab4d33/packages/middleware-signing/src/configurations.ts#L73) was probably based on assumption that the AWS service model may not have SigV4 trait, or it didn't exist back then. Given that now SigV4 trait is going to be present for every AWS service, we could just rely on that value instead of value from RegionInfoProvider(RIP). But that cleanup for AWS services, can be handled separately. I'm only going to change the logic for non-AWS service in this PR.

File: clients/client-accessanalyzer/AccessAnalyzerClient.ts [hunk 4]
```diff
@@ -226,7 +226,7 @@ export interface ClientDefaults extends Partial<__SmithyResolvedConfiguration<__
   serviceId?: string;
 
   /**
-   * The AWS region to which this client will send requests or use as signingRegion
+   * The AWS region to which this client will send requests
```
Comment (gosar) [inline clients/client-accessanalyzer/AccessAnalyzerClient.ts:229]: These are basically going back what the comment was before some of my earlier changes. See https://github.com/aws/aws-sdk-js-v3/blob/9ab71e8d3f404b69ff631bafb74eefab46d69309/clients/client-accessanalyzer/AccessAnalyzerClient.ts#L229

File: codegen/smithy-aws-typescript-codegen/src/main/java/software/amazon/smithy/aws/typescript/codegen/AddAwsRuntimeConfig.java (hunk 1 above)
Comment (JordonPhillips) [reply codegen/smithy-aws-typescript-codegen/src/main/java/software/amazon/smithy/aws/typescript/codegen/AddAwsRuntimeConfig.java:104]: Yes, but you can do that and still write the region separately. The methods in `TypeScriptWriter` return the object so you can chain calls like this, but it's not necessary.

File: codegen/smithy-aws-typescript-codegen/src/main/java/software/amazon/smithy/aws/typescript/codegen/EndpointGenerator.java (hunk 3 above)
Comment (JordonPhillips) [reply codegen/smithy-aws-typescript-codegen/src/main/java/software/amazon/smithy/aws/typescript/codegen/EndpointGenerator.java]: That's a fair question. It's probably better to just fail here rather than potentially silently failing

**Review by JordonPhillips — APPROVED** [review]

File: protocol_tests/aws-restjson/tests/functional/restjson1.spec.ts [hunk 5]
```diff
@@ -2352,8 +2352,8 @@ it("RestJsonInputAndOutputWithNumericHeaders:Response", async () => {
     ...clientParams,
     requestHandler: new ResponseDeserializationTestHandler(true, 200, {
       "x-float": "1.1",
-      "x-byte": "1",
       "x-long": "123",
+      "x-byte": "1",
```
Comment (AllanZhengYP) [inline protocol_tests/aws-restjson/tests/functional/restjson1.spec.ts:2356]: I don't think this change is needed.

File: protocol_tests/aws-restjson/tests/functional/restjson1.spec.ts [hunk 6]
```diff
@@ -5897,8 +5897,8 @@ it("RestJsonTimestampFormatHeaders:Response", async () => {
     requestHandler: new ResponseDeserializationTestHandler(true, 200, {
       "x-targetepochseconds": "1576540098",
       "x-memberdatetime": "2019-12-16T23:48:18Z",
-      "x-defaultformat": "Mon, 16 Dec 2019 23:48:18 GMT",
       "x-memberepochseconds": "1576540098",
+      "x-defaultformat": "Mon, 16 Dec 2019 23:48:18 GMT",
```
Comment (AllanZhengYP) [inline protocol_tests/aws-restjson/tests/functional/restjson1.spec.ts:5901]: This change is not needed.

File: protocol_tests/aws-restxml/tests/functional/restxml.spec.ts [hunk 7]
```diff
@@ -2315,8 +2315,8 @@ it("InputAndOutputWithNumericHeaders:Response", async () => {
       200,
       {
         "x-float": "1.1",
-        "x-byte": "1",
         "x-long": "123",
+        "x-byte": "1",
```
Comment (AllanZhengYP) [inline protocol_tests/aws-restxml/tests/functional/restxml.spec.ts:2319]: This change is not needed.

File: protocol_tests/aws-restxml/tests/functional/restxml.spec.ts [hunk 8]
```diff
@@ -3585,8 +3585,8 @@ it("TimestampFormatHeaders:Response", async () => {
       {
         "x-targetepochseconds": "1576540098",
         "x-memberdatetime": "2019-12-16T23:48:18Z",
-        "x-defaultformat": "Mon, 16 Dec 2019 23:48:18 GMT",
         "x-memberepochseconds": "1576540098",
+        "x-defaultformat": "Mon, 16 Dec 2019 23:48:18 GMT",
```
Comment (AllanZhengYP) [inline protocol_tests/aws-restxml/tests/functional/restxml.spec.ts:3589]: This change is not needed.

**Review by AllanZhengYP — APPROVED** [review]
It looks good to me! I only found some files changed unexpectly.

**Review by AllanZhengYP — APPROVED** [review]
